[ignore] [SymmMem] Add signal_op test with ring broadcast example

codingwithsurya opened the pull request:
**update: closing this as this test is duplicate, and we already have a test for putmem_signalblock + was flaky. we should look into adding a test with signal_op and signal_wait_until**

Previously, the `signal_op` test was hanging because it wasn't being used correctly with `signal_wait_until`. This test has the proper usage pattern by implementing a ring broadcast algorithm that mirrors the [NVSHMEM ring broadcast example](https://docs.nvidia.com/nvshmem/api/examples.html#ring-broadcast-example).
  
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #159701
* #159215
* #159136
* #158718
* __->__ #158676
* #158515

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

codingwithsurya committed:
add nvshmem signal op test w/ ring broadcast

[ghstack-poisoned]

codingwithsurya committed:
Update on "[SymmMem] Add signal_op test with ring broadcast example"

Previously, the `signal_op` test was hanging because it wasn't being used correctly with `signal_wait_until`. This test has the proper usage pattern by implementing a ring broadcast algorithm that mirrors the [NVSHMEM ring broadcast example](https://docs.nvidia.com/nvshmem/api/examples.html#ring-broadcast-example).
  

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

codingwithsurya committed:
Update on "[SymmMem] Add signal_op test with ring broadcast example"

Previously, the `signal_op` test was hanging because it wasn't being used correctly with `signal_wait_until`. This test has the proper usage pattern by implementing a ring broadcast algorithm that mirrors the [NVSHMEM ring broadcast example](https://docs.nvidia.com/nvshmem/api/examples.html#ring-broadcast-example).
  

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

fduwjj commented on test/distributed/test_nvshmem_triton.py:
```diff
@@ -667,6 +672,73 @@ def test_triton_signal_wait_until(self) -> None:
+                peer=next_pe,
+                extern_libs=nvshmem_lib,
+            )
+        dist.barrier()
```
what if we call another fence_kernel here? barrier will call an all-reduce inside.

codingwithsurya replied:
wouldn't nvshmem.barrier_all() be a better idea? nvshmem_fence() is basically a memory ordering primitive but here we want some type of synchronization to make sure we don't run into any race conditions. barrier_all is more of a collective synchronization primitive.

kwen2501 commented:
What would be a fix to the "previous" test?

codingwithsurya committed:
Update on "[SymmMem] Add signal_op test with ring broadcast example"

Previously, the `signal_op` test was hanging because it wasn't being used correctly with `signal_wait_until`. This test has the proper usage pattern by implementing a ring broadcast algorithm that mirrors the [NVSHMEM ring broadcast example](https://docs.nvidia.com/nvshmem/api/examples.html#ring-broadcast-example).
  

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]
